Classes/Private/Manager/Audio/OCTAudioQueue.m

```diff
@@ -0,0 +1,418 @@
+{
+    return [NSError errorWithDomain:NSOSStatusErrorDomain
+                               code:resultCode
+                           userInfo:@{NSLocalizedDescriptionKey : @"Consult the CoreAudio header files/google for the meaning of the error code."}];
+}
+
+#if ! TARGET_OS_IPHONE
+static NSString *OCTGetSystemAudioDevice(AudioObjectPropertySelector sel, NSError **err)
+{
+    AudioDeviceID devID = 0;
+    OSStatus ok = 0;
+    UInt32 size = sizeof(AudioDeviceID);
+    AudioObjectPropertyAddress address = {
+        .mSelector = sel,
+        .mScope = kAudioObjectPropertyScopeGlobal,
+        .mElement = kAudioObjectPropertyElementMaster
+    };
+
+    ok = _AudioObjectGetPropertyData(kAudioObjectSystemObject, &address, 0, NULL, &size, &devID);
+    if (ok != kAudioHardwareNoError) {
+        DDLogCError(@"failed AudioObjectGetPropertyData for system object: %d! Crash may or may not be imminent", ok);
+        if (err) {
+            *err = OCTErrorFromCoreAudioCode(ok);
+        }
+        return nil;
+    }
+
+    address.mSelector = kAudioDevicePropertyDeviceUID;
+    CFStringRef unique = NULL;
+    size = sizeof(unique);
+    ok = _AudioObjectGetPropertyData(devID, &address, 0, NULL, &size, &unique);
+    if (ok != kAudioHardwareNoError) {
+        DDLogCError(@"failed AudioObjectGetPropertyData for selected device: %d! Crash may or may not be imminent", ok);
+        if (err) {
+            *err = OCTErrorFromCoreAudioCode(ok);
+        }
+        return nil;
+    }
+
+    return (__bridge NSString *)unique;
+}
+#endif
+
+@interface OCTAudioQueue ()
+
+// use this to track what nil means in terms of audio device
+@property (assign, nonatomic) BOOL isOutput;
+@property (assign, nonatomic) AudioStreamBasicDescription streamFmt;
+@property (assign, nonatomic) AudioQueueRef audioQueue;
+@property (assign, nonatomic) TPCircularBuffer buffer;
+@property (assign, nonatomic) BOOL running;
+
+@end
+
+@implementation OCTAudioQueue {
+    AudioQueueBufferRef _AQBuffers[kNumberOfAudioQueueBuffers];
+}
+
+- (instancetype)initWithInputDeviceID:(NSString *)devID error:(NSError **)error
+{
+#if TARGET_OS_IPHONE
+    AVAudioSession *session = [AVAudioSession sharedInstance];
+    _streamFmt.mSampleRate = session.sampleRate;
+#else
+    _streamFmt.mSampleRate = kDefaultSampleRate;
+#endif
+    _streamFmt.mFormatID = kAudioFormatLinearPCM;
+    _streamFmt.mFormatFlags = kLinearPCMFormatFlagIsSignedInteger;
+    _streamFmt.mChannelsPerFrame = kNumberOfInputChannels;
+    _streamFmt.mBytesPerFrame = kBytesPerSample * kNumberOfInputChannels;
+    _streamFmt.mBitsPerChannel = kBitsPerByte * kBytesPerSample;
+    _streamFmt.mFramesPerPacket = kFramesPerPacket;
+    _streamFmt.mBytesPerPacket = kBytesPerSample * kNumberOfInputChannels * kFramesPerPacket;
+    _isOutput = NO;
+    _deviceID = devID;
+
+    TPCircularBufferInit(&_buffer, kBufferLength);
+    OSStatus res = 0;
+    if ((res = [self createAudioQueue]) != 0) {
```

Comment:
Instead of comparing it to `0` I think it would be better to compare it to `noErr`. Just to keep it consistent with the rest of the class.
```
OSStatus res = [self createAudioQueue];
if (res != noErr) {
```